### PR TITLE
CFE-815: Add externalCertificate to Route spec 

### PR DESCRIPTION
**Pull request by thejasn**
## Description
API changes required as part of https://github.com/openshift/enhancements/pull/1307. Introduces a new feature gate `RouteCertificateReference` as part of `TechPreviewNoUpgrade` feature set.

### CODE REVIEW

File: route/v1/types.go [hunk 1]
```diff
@@ -225,6 +229,44 @@ type RouteIngressCondition struct {
 	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
 }
 
+// RouteCertificateConditionType is a valid value forRouteCertificateCondition
+type RouteCertificateConditionType string
+
+// RouteCertificateCondition contains details of the certificate being served by routers associated with
+// this route.
+type RouteCertificateCondition struct {
```
Comment (JoelSpeed) [inline route/v1/types.go:237]: Apart from type checking on the `Type` field within this, is there any reason not to just use a `metav1.Condition`?

File: route/v1/types.go [hunk 2]
```diff
@@ -280,6 +280,15 @@ type TLSConfig struct {
 	// * Disable - no traffic is allowed on the insecure port.
 	// * Redirect - clients are redirected to the secure port.
 	InsecureEdgeTerminationPolicy InsecureEdgeTerminationPolicyType `json:"insecureEdgeTerminationPolicy,omitempty" protobuf:"bytes,6,opt,name=insecureEdgeTerminationPolicy,casttype=InsecureEdgeTerminationPolicyType"`
+
+	// certificateRef provides certificate contents as a secret reference.
```
[the remark this answers is not in the document]
Comment (thejasn) [reply route/v1/types.go:289]: Kinda gets duplicated with the godoc on `LocalObjectReference` (now that it is defined locally). Let me know if that is the case.

File: route/v1/types.go [hunk 3]
```diff
@@ -284,6 +284,26 @@ type TLSConfig struct {
...
+	// This should be a single serving certificate, not a certificate
+	// chain. Do not include a CA certificate. The secret referenced should
+	// be present in the same namespace as that of the Route. If certificateRef
+	// provided, do not set certificate.
+	//
+	// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+	// +optional
+	CertificateRef LocalObjectReference `json:"certificateRef,omitempty" protobuf:"bytes,7,opt,name=certificateRef"`
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply route/v1/types.go:296]: +1 to the comments here, `ExternalCertificate` is better than `certificateRef` IMO

File: route/v1/types.go [hunk 4]
```diff
@@ -284,6 +284,26 @@ type TLSConfig struct {
...
+}
+
+// LocalObjectReference contains enough information to let you locate the
+// referenced object inside the same namespace.
+// +structType=atomic
+type LocalObjectReference struct {
+	// Name of the referent.
```
Comment (JoelSpeed) [inline route/v1/types.go:303]: `Name` should be `name`

File: route/v1/types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply route/v1/types.go:288]: The godoc on the struct there does not get copied across to generated docs, so should include it in this godoc section as well

File: route/v1/types.go [hunk 5]
```diff
@@ -248,6 +248,7 @@ type RouterShard struct {
 // TLSConfig defines config used to secure a route and provide termination
 //
 // +kubebuilder:validation:XValidation:rule="has(self.termination) && has(self.insecureEdgeTerminationPolicy) ? !((self.termination=='passthrough') && (self.insecureEdgeTerminationPolicy=='Allow')) : true", message="cannot have both spec.tls.termination: passthrough and spec.tls.insecureEdgeTerminationPolicy: Allow"
+// +openshift:validation:FeatureSetAwareXValidation:featureSet=TechPreviewNoUpgrade,rule="!(has(self.certificate) && has(self.externalCertificate))", message="cannot have both spec.tls.certificate and spec.tls.externalCertificate"
```
Comment (JoelSpeed) [inline route/v1/types.go:251]: You'll want to include a `CustomNoUpgrade` for this too else the feature gate could not be built on a custom cluster
```suggestion
// +openshift:validation:FeatureSetAwareXValidation:featureSet=TechPreviewNoUpgrade;CustomNoUpgrade,rule="!(has(self.certificate) && has(self.externalCertificate))", message="cannot have both spec.tls.certificate and spec.tls.externalCertificate"
```

File: route/v1/types.go [hunk 6]
```diff
@@ -284,6 +285,26 @@ type TLSConfig struct {
 	//
 	// +kubebuilder:validation:Enum=Allow;None;Redirect;""
 	InsecureEdgeTerminationPolicy InsecureEdgeTerminationPolicyType `json:"insecureEdgeTerminationPolicy,omitempty" protobuf:"bytes,6,opt,name=insecureEdgeTerminationPolicy,casttype=InsecureEdgeTerminationPolicyType"`
+
+	// externalCertificate provides certificate contents as a secret reference.
+	// This should be a single serving certificate, not a certificate
+	// chain. Do not include a CA certificate. The secret referenced should
+	// be present in the same namespace as that of the Route. If externalCertificate is
+	// provided, do not set certificate.
```
Comment (JoelSpeed) [inline route/v1/types.go:293]: I would make this a stronger requirement here
```suggestion
	// be present in the same namespace as that of the Route.
	// Forbidden when `certificate` is set.
```

**Review by JoelSpeed — COMMENTED** [review]
Couple of minor things but otherwise I think LGTM

File: route/v1/types.go (hunk 6 above)
Comment (thejasn) [reply route/v1/types.go]: Addressed

**Comment by thejasn** [comment]
Added test suite for `CustomNoUpgrade`

**Comment by ingvagabund** [comment]
Merging this PR without updating OA code caused this trouble: https://issues.redhat.com/browse/OCPBUGS-16264. In general, every PR changing any type should have a corresponding OA PR to avoid similar troubleshooting in the future.

**Comment by thejasn** [comment]
> Merging this PR without updating OA code caused this trouble: https://issues.redhat.com/browse/OCPBUGS-16264. In general, every PR changing any type should have a corresponding OA PR to avoid similar troubleshooting in the future.

Hey @ingvagabund,
Sorry, I was not aware of this potential failure point due to the conversion generator in the apiserver. Will keep this in mind going forward.

**Comment by ingvagabund** [comment]
No worries :) I did not expect this either.

**Comment by JoelSpeed** [comment]
Given the enhancement for this hasn't merged yet, we probably shouldn't have merged the API either. Should we revert this for now until the enhancement has had an OK? (especially given this isn't ready for 4.14 with branch cut less than two weeks away)

What are the implications for openshift-apiserver if we revert this @ingvagabund ?

**Pull request merged**